### ISSUE TITLE
Is it appropriate for highly imbalanced data?

### ISSUE THREAD

**Issue by thehemen** [KevinMusgrave/pytorch-metric-learning#238]
Hello.
I have a dataset of 22000 samples and 3500 classes. One class has about 700 samples but the majority of them is 6 samples only.
I'm trying to implemented the metric learning but the results are poor. It gives only 0.08 "precision_at_1" after 25 epochs of training with the scheduler.

```python3
m = 6
batch_size = 192
num_epochs = 25

train_sampler = samplers.MPerClassSampler(y_train, m=m, batch_size=batch_size)
train_loader = torch.utils.data.DataLoader(train_dataset, batch_size=batch_size, sampler=train_sampler,
                                           num_workers=4)
test_loader = torch.utils.data.DataLoader(test_dataset, batch_size=batch_size)

distance = distances.CosineSimilarity()
reducer = reducers.AvgNonZeroReducer()
loss_func = losses.ContrastiveLoss(pos_margin=1, neg_margin=0, reducer = reducer, distance = distance)
mining_func = miners.MultiSimilarityMiner(epsilon=0.1, distance = distance)
accuracy_calculator = AccuracyCalculator(include = (accuracy_type,), avg_of_avgs=True)

optimizer = optim.Adam(model.parameters(), lr=1e-3, weight_decay=1e-5)
scheduler = optim.lr_scheduler.OneCycleLR(optimizer=optimizer,
                                          steps_per_epoch=len(train_loader),
                                          pct_start=0.1,
                                          div_factor=1e3, 
                                          max_lr=1e-3,
                                          epochs=num_epochs)
```

Is it possible to improve the results somehow? Is it appropriate for such an imbalanced dataset?

**Comment by KevinMusgrave** [comment]
Is that 0.08 on the training set or validation set?

**Comment by KevinMusgrave** [comment]
You can also try logging various things to see how training is progressing:

- ```loss_func.reducer.reducers["pos_loss"].pos_pairs_past_filter```
- ```loss_func.reducer.reducers["neg_loss"].neg_pairs_past_filter```
- ```miner.num_pos_pairs```
- ```miner.num_neg_pairs```

You can try visualizing the embedding space to see if the dominant class is separated from the few-shot classes. Here's a function you can use for that (it requires [UMAP](https://github.com/lmcinnes/umap) which you can install with ```pip install umap-learn```).

```python
import umap
import numpy as np
import matplotlib.pyplot as plt
from cycler import cycler

def plot_umap(embeddings, labels):
    umap_embeddings = umap.UMAP().fit_transform(embeddings)
    label_set = np.unique(labels)
    num_classes = len(label_set)
    fig = plt.figure(figsize=(20,15))
    plt.gca().set_prop_cycle(cycler("color", [plt.cm.nipy_spectral(i) for i in np.linspace(0, 0.9, num_classes)]))
    for i in range(num_classes):
        idx = labels == label_set[i]
        plt.plot(umap_embeddings[idx, 0], umap_embeddings[idx, 1], ".", markersize=1)   
    plt.show()
```

**Comment by thehemen** [comment]
I've added these metrics and now it looks like this. According to the train accuracy, it's really overfitting...

<details>
<summary>Training logs...</summary>
Seed: 0, fold: 0, epoch: 00: train acc = 0.113688, test acc = 0.043992, num_pos_pairs: 674, num_neg_pairs: 5415.
pos_pairs_past_filter: 666, neg_pairs_past_filter: 4908.


Seed: 0, fold: 0, epoch: 01: train acc = 0.194674, test acc = 0.047328, num_pos_pairs: 629, num_neg_pairs: 5370.
pos_pairs_past_filter: 629, neg_pairs_past_filter: 4286.

Seed: 0, fold: 0, epoch: 02: train acc = 0.288983, test acc = 0.062481, num_pos_pairs: 363, num_neg_pairs: 2623.
pos_pairs_past_filter: 363, neg_pairs_past_filter: 1899.

Seed: 0, fold: 0, epoch: 03: train acc = 0.375784, test acc = 0.071251, num_pos_pairs: 322, num_neg_pairs: 635.
pos_pairs_past_filter: 320, neg_pairs_past_filter: 635.

Seed: 0, fold: 0, epoch: 04: train acc = 0.434029, test acc = 0.066375, num_pos_pairs: 146, num_neg_pairs: 282.
pos_pairs_past_filter: 146, neg_pairs_past_filter: 282.

Seed: 0, fold: 0, epoch: 05: train acc = 0.512191, test acc = 0.071995, num_pos_pairs: 217, num_neg_pairs: 421.
pos_pairs_past_filter: 217, neg_pairs_past_filter: 421.

Seed: 0, fold: 0, epoch: 06: train acc = 0.566040, test acc = 0.062668, num_pos_pairs: 162, num_neg_pairs: 324.
pos_pairs_past_filter: 162, neg_pairs_past_filter: 324.

Seed: 0, fold: 0, epoch: 07: train acc = 0.600001, test acc = 0.065036, num_pos_pairs: 85, num_neg_pairs: 103.
pos_pairs_past_filter: 85, neg_pairs_past_filter: 103.

Seed: 0, fold: 0, epoch: 08: train acc = 0.661683, test acc = 0.061660, num_pos_pairs: 98, num_neg_pairs: 153.
pos_pairs_past_filter: 98, neg_pairs_past_filter: 153.

Seed: 0, fold: 0, epoch: 09: train acc = 0.685253, test acc = 0.061174, num_pos_pairs: 165, num_neg_pairs: 244.
pos_pairs_past_filter: 165, neg_pairs_past_filter: 244.

Seed: 0, fold: 0, epoch: 10: train acc = 0.724406, test acc = 0.062457, num_pos_pairs: 104, num_neg_pairs: 281.
pos_pairs_past_filter: 104, neg_pairs_past_filter: 281.

Seed: 0, fold: 0, epoch: 11: train acc = 0.765203, test acc = 0.064708, num_pos_pairs: 106, num_neg_pairs: 205.
pos_pairs_past_filter: 106, neg_pairs_past_filter: 205.

Seed: 0, fold: 0, epoch: 12: train acc = 0.794025, test acc = 0.065459, num_pos_pairs: 27, num_neg_pairs: 54.
pos_pairs_past_filter: 27, neg_pairs_past_filter: 54.

Seed: 0, fold: 0, epoch: 13: train acc = 0.819646, test acc = 0.062672, num_pos_pairs: 25, num_neg_pairs: 25.
pos_pairs_past_filter: 25, neg_pairs_past_filter: 25.

Seed: 0, fold: 0, epoch: 14: train acc = 0.834340, test acc = 0.073535, num_pos_pairs: 27, num_neg_pairs: 38.
pos_pairs_past_filter: 27, neg_pairs_past_filter: 38.

Seed: 0, fold: 0, epoch: 15: train acc = 0.866927, test acc = 0.073514, num_pos_pairs: 7, num_neg_pairs: 5.
pos_pairs_past_filter: 7, neg_pairs_past_filter: 5.

Seed: 0, fold: 0, epoch: 16: train acc = 0.890524, test acc = 0.070340, num_pos_pairs: 56, num_neg_pairs: 64.
pos_pairs_past_filter: 56, neg_pairs_past_filter: 64.

Seed: 0, fold: 0, epoch: 17: train acc = 0.906599, test acc = 0.073651, num_pos_pairs: 16, num_neg_pairs: 16.
pos_pairs_past_filter: 16, neg_pairs_past_filter: 16.

Seed: 0, fold: 0, epoch: 18: train acc = 0.915844, test acc = 0.070955, num_pos_pairs: 0, num_neg_pairs: 0.
pos_pairs_past_filter: 0, neg_pairs_past_filter: 0.

Seed: 0, fold: 0, epoch: 19: train acc = 0.929157, test acc = 0.075838, num_pos_pairs: 53, num_neg_pairs: 52.
pos_pairs_past_filter: 53, neg_pairs_past_filter: 52.

Seed: 0, fold: 0, epoch: 20: train acc = 0.937617, test acc = 0.076362, num_pos_pairs: 65, num_neg_pairs: 72.
pos_pairs_past_filter: 65, neg_pairs_past_filter: 72.

Seed: 0, fold: 0, epoch: 21: train acc = 0.942302, test acc = 0.080954, num_pos_pairs: 12, num_neg_pairs: 24.
pos_pairs_past_filter: 12, neg_pairs_past_filter: 24.

Seed: 0, fold: 0, epoch: 22: train acc = 0.946324, test acc = 0.079844, num_pos_pairs: 0, num_neg_pairs: 0.
pos_pairs_past_filter: 0, neg_pairs_past_filter: 0.

Seed: 0, fold: 0, epoch: 23: train acc = 0.947880, test acc = 0.081208, num_pos_pairs: 0, num_neg_pairs: 0.
pos_pairs_past_filter: 0, neg_pairs_past_filter: 0.

Seed: 0, fold: 0, epoch: 24: train acc = 0.948272, test acc = 0.080541, num_pos_pairs: 0, num_neg_pairs: 0. 
pos_pairs_past_filter: 0, neg_pairs_past_filter: 0.
</details>

I've forgotten to mention that there're a few other classes less than 700 but more 100 samples. So the plot is.
![Figure 1](https://user-images.githubusercontent.com/12379560/100193766-02ac2c00-2f27-11eb-9a62-a23c46e77631.png)

**Comment by KevinMusgrave** [comment]
Is that the plot for the training or test set?

Based on your training logs, I guess technically it isn't overfitting because the test set accuracy actually keeps going up. Do the training and test sets have the same classes and class imbalances?

**Comment by thehemen** [comment]
> Is that the plot for the training or test set?

It's from test set.

> Based on your training logs, it might not be overfitting because the test set accuracy actually keeps going up. Do the training and test sets have the same classes and class imbalances?

I'm using StratifiedKFold from scikit-learn to split by 80% / 20% so both train and test sets should have the same classes.

```python3
from sklearn.model_selection import StratifiedKFold

SEEDS = 5
NFOLDS = 5

for seed_id in range(SEEDS):
    kfold_col = f'kfold_{seed_id}'
    kf = StratifiedKFold(n_splits=NFOLDS, shuffle=True, random_state=seed_id)

    for f, (t_idx, v_idx) in enumerate(kf.split(X=train[feature_cols], y=train[target_col])):
        train.loc[v_idx, kfold_col] = int(f)

    train[kfold_col] = train[kfold_col].astype(int)
```

**Comment by KevinMusgrave** [comment]
Is it 6 samples / class in the validation set, or 6 samples per class for the complete dataset?

If it's 6 samples / class for the complete dataset, then 20% of 6 is 1.2, which is probably being rounded down to 1 (not rounded up to 2.) Therefore all the few-shot classes in the validation set will never have accuracy greater than 0.

**Comment by thehemen** [comment]
> Is it 6 samples / class in the validation set, or 6 samples per class for the complete dataset?

It is 6 samples per class in the complete dataset.

> If it's 6 samples / class for the complete dataset, then 20% of 6 is 1.2, which is probably being rounded down to 1 (not rounded up to 2.) Therefore all the few-shot classes in the validation set will never have accuracy greater than 0.

Hmm, that's an interesting observation that I had to keep in mind... Well, I'll try to use 3 folds only (so that it gives 2 samples for the validation set) to write down the results a bit later.

**Comment by thehemen** [comment]
> If it's 6 samples / class for the complete dataset, then 20% of 6 is 1.2, which is probably being rounded down to 1 (not rounded up to 2.) Therefore all the few-shot classes in the validation set will never have accuracy greater than 0.

I've investigated the issue with the number of samples in valid set when choosing 5 folds, and it's appeared that it's not rounded completely. Two samples may appear in any fold.

![Снимок экрана от 2020-11-26 11-59-38](https://user-images.githubusercontent.com/12379560/100310370-9c371480-2fdf-11eb-882c-9021c71bfcb6.png)
